clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -331,31 +332,59 @@ else if (error.exception() instanceof RetriableException)
      */
     private void resetOffset(TopicPartition partition) {
         OffsetResetStrategy strategy = subscriptions.resetStrategy(partition);
-        final long timestamp;
-        if (strategy == OffsetResetStrategy.EARLIEST)
-            timestamp = ListOffsetRequest.EARLIEST_TIMESTAMP;
-        else if (strategy == OffsetResetStrategy.LATEST)
-            timestamp = ListOffsetRequest.LATEST_TIMESTAMP;
-        else
-            throw new NoOffsetForPartitionException(partition);
-
         log.debug("Resetting offset for partition {} to {} offset.", partition, strategy.name().toLowerCase(Locale.ROOT));
-        long offset = getOffsetsByTimes(Collections.singletonMap(partition, timestamp), Long.MAX_VALUE).get(partition).offset();
-
+        long offset = listOffset(partition, strategy);
         // we might lose the assignment while fetching the offset, so check it is still active
         if (subscriptions.isAssigned(partition))
             this.subscriptions.seek(partition, offset);
     }
 
+    private long listOffset(TopicPartition partition, OffsetResetStrategy strategy) {
+        final long timestamp;
+        switch (strategy) {
+            case EARLIEST:
+                timestamp = ListOffsetRequest.EARLIEST_TIMESTAMP;
+                break;
+            case LATEST:
+                timestamp = ListOffsetRequest.LATEST_TIMESTAMP;
+                break;
+            default:
+                throw new NoOffsetForPartitionException(partition);
+        }
+        while (true) {
+            RequestFuture<Map<TopicPartition, OffsetAndTimestamp>> future =
+                    sendListOffsetRequests(false, Collections.singletonMap(partition, timestamp));
+            client.poll(future);
+            if (future.succeeded()) {
+                Map<TopicPartition, OffsetAndTimestamp> ret = future.value();
+                return ret.get(partition).offset();
```

Comment:
We didn't have it before, but maybe we should add a null check here for more resilience in the future.